src/main/java/net/unicon/lti/controller/lti/LTI3Controller.java

```diff
@@ -64,19 +69,105 @@ public class LTI3Controller {
     @Autowired
     LTIDataService ltiDataService;
 
+    @Autowired
+    DeepLinkService deepLinkService;
+
     @Autowired
     LtiContextRepository ltiContextRepository;
 
     @RequestMapping({"", "/"})
     public String lti3(HttpServletRequest req, Model model) throws DataServiceException, ConnectionException {
 
-        //First we will get the state, validate it
+        //We need to pass to the model the LTI3 request, the state and the nonce and link, and a JWT with the hash of all of them.
+        String state = req.getParameter("state");
+        String id_token = req.getParameter("id_token");
+
+
+        NonceState nonceState = ltiDataService.getRepos().nonceStateRepository.findByStateHash(state);
+        if (nonceState != null) {
+            Jws<Claims> stateClaims = ltijwtService.validateState(nonceState.getState());
+            String nonce = stateClaims.getBody().getId();
+            String tohash = id_token + state + nonce;
+            String expected_hash = Hashing.sha256()
+                    .hashString(tohash, StandardCharsets.UTF_8)
+                    .toString();
+            model.addAttribute("expected_state", state);
+            model.addAttribute("expected_nonce", nonce);
+            try {
+                String token = ltijwtService.generateStateNonceTokenJWT(expected_hash);
+                model.addAttribute("token", token);
+            } catch (GeneralSecurityException | IOException e) {
+                log.error("Error generating JWT", e);
```

Comment:
Could you add more detail to this error message, like "Error generating state nonce token JWT"?